include/deal.II/base/lazy.h

```diff
@@ -76,8 +76,13 @@ DEAL_II_NAMESPACE_OPEN
  *   Lazy<FullMatrix<double>> prolongation_matrix;
  * };
  * ```
+ *
+ * @dealiiConceptRequires{std::is_constructible_v<T, T> &&
+                          std::is_assignable_v<T &, T &&>}
  */
 template <typename T>
+DEAL_II_CXX20_REQUIRES((std::is_constructible_v<T, T> &&
+                        std::is_assignable_v<T &, T &&>))
```

Comment:
Do you mean
```suggestion
DEAL_II_CXX20_REQUIRES((std::is_copy_constructible_v<T> &&
                        std::is_move_assignable_v<T>))
```
?

Reply:
No, I actually meant what I had. The error I'm trying to prevent is in a statement of the form
```
  object = std::move(callable());
```
which pertains to case (4) here: https://en.cppreference.com/w/cpp/utility/optional/operator%3D . The `requires` clause exactly matches the documentation of what case (4) requires. It is possible that one can replace `std::is_constructible_v<T, T>` by `std::is_copy_constructible_v<T>` (see the "Possible implementation" case at https://en.cppreference.com/w/cpp/types/is_copy_constructible ) but I don't think it adds clarity to add all of the qualifications shown there.

Reply:
Fair enough. The documentation there states
> The function does not participate in overload resolution unless [std::decay_t](http://en.cppreference.com/w/cpp/types/decay)<U>(until C++20)[std::remove_cvref_t](http://en.cppreference.com/w/cpp/types/remove_cvref)<U>(since C++20) is not [std::optional](http://en.cppreference.com/w/cpp/utility/optional)<T>, [std::is_constructible_v](http://en.cppreference.com/w/cpp/types/is_constructible)<T, U> is true, [std::is_assignable_v](http://en.cppreference.com/w/cpp/types/is_assignable)<T&, U> is true, and at least one of the following is true:
>   - T is not a [scalar type](https://en.cppreference.com/w/cpp/language/type);
>   - [std::decay_t](http://en.cppreference.com/w/cpp/types/decay)<U> is not T.

So shouldn't we at least use `std::is_assignable_v<T&, T>` instead of `std::is_assignable_v<T&, T&&>`?

Reply:
I can change that tonight (when I'm back at my laptop), but does it actually make a difference? If you can assign with a *value* of type `U` on the right hand side, you can also assign if the right hand side is an rvalue reference to an object of type `U` and vice versa. Or do I misremember? The only corner case I can think of is a situation like
```
  x = 1;
```
where you have a value on the rhs, but I'm unsure whether you can have an rvalue reference. Such literal values have become more common with user-defined literals.

Reply:
Playing a little bit with this (https://godbolt.org/z/zWz9fcG9e), it really doesn't make a difference and it seems the behavior we want is
```
std::is_move_constructible_v<T> && std::is_move_assignable_v<T>
```
which still is better understandable in my opinion.